android/jni/com/mapswithme/maps/uber/Uber.cpp

```diff
@@ -0,0 +1,99 @@
+#include "../Framework.hpp"
+
+#include "../../core/jni_helper.hpp"
+#include "partners_api/uber_api.hpp"
+
+namespace
+{
+jclass g_uberClass;
+jclass g_productClass;
+jclass g_routingControllerClass;
+jclass g_uberInfoClass;
+jmethodID g_uberInfoConstructor;
+jobject g_routingControllerInstance;
+jmethodID g_productConstructor;
+jmethodID g_routingControllerGetMethod;
+jmethodID g_uberInfoCallbackMethod;
+jclass g_uberLinksClass;
+jmethodID g_uberLinksConstructor;
+jlong lastRequestId;
+
+void PrepareClassRefs(JNIEnv * env)
+{
+  if (g_uberClass)
+    return;
+
+  g_uberClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/uber/UberInfo");
+  g_productClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/uber/UberInfo$Product");
+  g_routingControllerClass =
+      jni::GetGlobalClassRef(env, "com/mapswithme/maps/routing/RoutingController");
+  g_productConstructor = jni::GetConstructorID(
+      env, g_productClass,
+      "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V");
+  g_uberInfoClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/uber/UberInfo");
+  g_routingControllerGetMethod = jni::GetStaticMethodID(
+      env, g_routingControllerClass, "get", "()Lcom/mapswithme/maps/routing/RoutingController;");
+  g_routingControllerInstance =
+      env->CallStaticObjectMethod(g_routingControllerClass, g_routingControllerGetMethod);
+  g_uberInfoCallbackMethod =
+      jni::GetMethodID(env, g_routingControllerInstance, "onUberInfoReceived",
+                       "(Lcom/mapswithme/maps/uber/UberInfo;)V");
+  g_uberInfoConstructor = jni::GetConstructorID(env, g_uberInfoClass,
+                                                "([Lcom/mapswithme/maps/uber/UberInfo$Product;)V");
+  g_uberLinksClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/uber/UberLinks");
+  g_uberLinksConstructor =
+      jni::GetConstructorID(env, g_uberLinksClass, "(Ljava/lang/String;Ljava/lang/String;)V");
+}
+}  // namespace
+
+extern "C" {
+
+JNIEXPORT void JNICALL Java_com_mapswithme_maps_uber_Uber_nativeRequestUberProducts(
+    JNIEnv * env, jclass clazz, jdouble srcLat, jdouble srcLon, jdouble dstLat, jdouble dstLon)
+{
+  PrepareClassRefs(env);
+
+  ms::LatLon const from(srcLat, srcLon);
+  ms::LatLon const to(dstLat, dstLon);
+
+  lastRequestId = static_cast<jlong>(g_framework->RequestUberProducts(
```

Comment:
I used the jlong, because I thought that it's required to be able to access the variable from JNI method, but I was wrong. My assumption is correct only for object references such as jobjec, jclass, jmethod. Fixed.